File: Examples/CSharp/Exchange/DownloadMessagesFromExchangeServerFoldersRecursively.cs
using System;
using System.IO;
using System.Net;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;
using Aspose.Email.Exchange;
using Aspose.Email.Mail;
using Aspose.Email.Outlook;

/*
This project uses Automatic Package Restore feature of NuGet to resolve Aspose.Email for .NET API reference 
when the project is build. Please check https:// Docs.nuget.org/consume/nuget-faq for more information. 
If you do not wish to use NuGet, you can manually download Aspose.Email for .NET API from http://www.aspose.com/downloads, 
install it and then add its reference to this project. For any issues, questions or suggestions 
please feel free to contact us using http://www.aspose.com/community/forums/default.aspx
*/

namespace Aspose.Email.Examples.CSharp.Email.Exchange
{
    class DownloadMessagesFromExchangeServerFoldersRecursively
    {
        static string mailboxURI = "https:// Ex2010/ews/exchange.asmx"; // EWS

        static string username = "administrator";
        static string password = "pwd";
        static string domain = "ex2010.local";

        public static void Run()
        {
            // ExStart:DeleteMessagesFromExchangeServer
            // Create instance of IEWSClient class by giving credentials
            string mailboxURI = "http:// Ex2003/exchange/administrator"; // WebDAV 

            // Register callback method for SSL validation event
            ServicePointManager.ServerCertificateValidationCallback += RemoteCertificateValidationHandler;

            try
            {

                DownloadAllMessages();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            // ExEnd:DeleteMessagesFromExchangeServer
        }

        private static void DownloadAllMessages()
        {
            try
            {
                string rootFolder = domain + "-" + username;
                Directory.CreateDirectory(rootFolder);
                string inboxFolder = rootFolder + "\\Inbox";
                Directory.CreateDirectory(inboxFolder);

                Console.WriteLine("Downloading all messages from Inbox....");
                // Create instance of IEWSClient class by giving credentials
                IEWSClient client = EWSClient.GetEWSClient("https://outlook.office365.com/ews/exchange.asmx", username, password, domain);

                ExchangeMailboxInfo mailboxInfo = client.GetMailboxInfo();
                Console.WriteLine("Mailbox URI: " + mailboxInfo.MailboxUri);
                string rootUri = client.GetMailboxInfo().RootUri;
                // List all the folders from Exchange server
                ExchangeFolderInfoCollection folderInfoCollection = client.ListSubFolders(rootUri);
                foreach (ExchangeFolderInfo folderInfo in folderInfoCollection)
                {
                    // Call the recursive method to read messages and get sub-folders
                    ListMessagesInFolder(client, folderInfo, rootFolder);
                }

                Console.WriteLine("All messages downloaded.");
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        /// <summary>
        /// Recursive method to get messages from folders and sub-folders
        /// </summary>
        /// <param name="folderInfo"></param>
        /// <param name="rootFolder"></param>
        private static void ListMessagesInFolder(IEWSClient client, ExchangeFolderInfo folderInfo, string rootFolder)
        {
            // Create the folder in disk (same name as on IMAP server)
            string currentFolder = rootFolder + "\\" + folderInfo.DisplayName;
            Directory.CreateDirectory(currentFolder);

            // List messages
            ExchangeMessageInfoCollection msgInfoColl = client.ListMessages(folderInfo.Uri);
            Console.WriteLine("Listing messages....");
            int i = 0;
            foreach (ExchangeMessageInfo msgInfo in msgInfoColl)
            {
                // Get subject and other properties of the message
                Console.WriteLine("Subject: " + msgInfo.Subject);

                // Get rid of characters like ? and :, which should not be included in a file name
                string fileName = msgInfo.Subject.Replace(":", " ").Replace("?", " ");

                // Save the message in MSG format
                MailMessage msg = client.FetchMessage(msgInfo.UniqueUri);
                // Msg.Save(msg);
                i++;
            }
            Console.WriteLine("============================\n");
            try
            {
                // If this folder has sub-folders, call this method recursively to get messages
                ExchangeFolderInfoCollection folderInfoCollection = client.ListSubFolders(folderInfo.Uri);
                foreach (ExchangeFolderInfo subfolderInfo in folderInfoCollection)
                {
                    ListMessagesInFolder(client, subfolderInfo, currentFolder);
                }
            }
            catch (Exception)
            {
            }
        }

        private static bool RemoteCertificateValidationHandler(object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors)
        {
            return true; // Ignore the checks and go ahead
        }
    }
}

File: Examples/VisualBasic/Exchange/DownloadMessagesFromExchangeServerFoldersRecursively.vb
Imports System.IO
Imports Aspose.Email.Exchange
Imports Aspose.Email.Mail
Imports Aspose.Email.Outlook

'
'This project uses Automatic Package Restore feature of NuGet to resolve Aspose.Email for .NET API reference 
'when the project is build. Please check https://Docs.nuget.org/consume/nuget-faq for more information. 
'If you do not wish to use NuGet, you can manually download Aspose.Email for .NET API from http://www.aspose.com/downloads, 
'install it and then add its reference to this project. For any issues, questions or suggestions 
'please feel free to contact us using http://www.aspose.com/community/forums/default.aspx
'


Namespace Aspose.Email.Examples.VisualBasic.Email.Exchange
    Class DownloadMessagesFromExchangeServerFoldersRecursively
        ' ExStart:DownloadMessagesFromExchangeServerFoldersRecursively
        Private Shared mailboxURI As String = "https://ex2010/ews/exchange.asmx" ' EWS
        Private Shared username As String = "administrator"
        Private Shared password As String = "pwd"
        Private Shared domain As String = "ex2010.local"

        Shared Sub Run()
            Try
                DownloadAllMessages()
            Catch ex As Exception
                Console.WriteLine(ex.Message)
            End Try
            ' ExEnd:DownloadMessagesFromExchangeServerFoldersRecursively
        End Sub

        ' ExStart:DownloadAllMessages
        Private Shared Sub DownloadAllMessages()
            Try
                Dim rootFolder As String = domain & "-" & username
                Directory.CreateDirectory(rootFolder)
                Dim inboxFolder As String = rootFolder & "\Inbox"
                Directory.CreateDirectory(inboxFolder)

                Console.WriteLine("Downloading all messages from Inbox....")
                ' Create instance of EWSClient class by giving credentials
                Dim client As IEWSClient = EWSClient.GetEWSClient("https://outlook.office365.com/ews/exchange.asmx", "testUser", "pwd", "domain")

                Dim mailboxInfo As ExchangeMailboxInfo = client.GetMailboxInfo()
                Console.WriteLine("Mailbox URI: " & mailboxInfo.MailboxUri)
                Dim rootUri As String = client.GetMailboxInfo().RootUri
                ' List all the folders from Exchange server
                Dim folderInfoCollection As ExchangeFolderInfoCollection = client.ListSubFolders(rootUri)
                For Each folderInfo As ExchangeFolderInfo In folderInfoCollection
                    ' Call the recursive method to read messages and get sub-folders
                    ListMessagesInFolder(client, folderInfo, rootFolder)
                Next folderInfo

                Console.WriteLine("All messages downloaded.")
            Catch ex As Exception
                Console.WriteLine(ex.Message)
            End Try
        End Sub

        ' Recursive method to get messages from folders and sub-folders
        Private Shared Sub ListMessagesInFolder(ByVal client As IEWSClient, ByVal folderInfo As ExchangeFolderInfo, ByVal rootFolder As String)
            ' Create the folder in disk (same name as on IMAP server)
            Dim currentFolder As String = rootFolder & "\" & folderInfo.DisplayName
            Directory.CreateDirectory(currentFolder)

            ' List messages
            Dim msgInfoColl As ExchangeMessageInfoCollection = client.ListMessages(folderInfo.Uri)
            Console.WriteLine("Listing messages....")
            Dim i As Integer = 0
            For Each msgInfo As ExchangeMessageInfo In msgInfoColl
                ' Get subject and other properties of the message
                Console.WriteLine("Subject: " & msgInfo.Subject)

                ' Get rid of characters like ? and :, which should not be included in a file name
                Dim fileName As String = msgInfo.Subject.Replace(":", " ").Replace("?", " ")

                ' Save the message in MSG format
                Dim msg As MailMessage = client.FetchMessage(msgInfo.UniqueUri)
                'msg.Save(currentFolder & "\" & fileName & "-" & i & ".msg", MailMessageSaveType.OutlookMessageFormat)
                i += 1
            Next msgInfo
            Console.WriteLine("============================" & Constants.vbLf)
            Try
                ' If this folder has sub-folders, call this method recursively to get messages
                Dim folderInfoCollection As ExchangeFolderInfoCollection = client.ListSubFolders(folderInfo.Uri)
                For Each subfolderInfo As ExchangeFolderInfo In folderInfoCollection
                    ListMessagesInFolder(client, subfolderInfo, currentFolder)
                Next subfolderInfo
            Catch e1 As Exception
            End Try
        End Sub
    End Class
    ' ExEnd:DownloadAllMessages
End Namespace
